src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java

```diff
@@ -0,0 +1,116 @@
+            currentReadsWithSameUMI = new ReadsWithSameUMI(read);
+            return;
+        }
+
+        final int readMoleculeNumber = MoleculeID.getMoleculeNumberOfRead(read);
+        final int duplicateSetMoleculeNumber = currentReadsWithSameUMI.getMoleculeNumber();
+
+        // If the incoming read has the molecule id less than that of the currentDuplicateSet,
+        // the input bam is not sorted properly by the MI tag
+        if (duplicateSetMoleculeNumber > readMoleculeNumber){
+            throw new UserException(String.format("The input bam must be sorted by the molecule ID (%s) tag.", SAMTag.MI.name()));
+        }
+
+        if (duplicateSetMoleculeNumber < readMoleculeNumber) {
+            // The incoming read's molecule ID does not match that of the current duplicate, meaning we've reached the end of the current set.
+            // Call the apply() method to process the current set and start a new set.
+            if (rejectSet(currentReadsWithSameUMI)){
+                currentReadsWithSameUMI = new ReadsWithSameUMI(read);
+                return;
+            }
+
+            apply(currentReadsWithSameUMI,
+                    new ReferenceContext(reference, currentReadsWithSameUMI.getInterval()), // Will create an empty ReferenceContext if reference or readInterval == null
+                    new FeatureContext(features, currentReadsWithSameUMI.getInterval()));
+            currentReadsWithSameUMI = new ReadsWithSameUMI(read);
+            return;
+        }
+
+        // The incoming read has the same UMI as the current set; simply add to the current set
+        currentReadsWithSameUMI.addRead(read);
+    }
+
+    /**
+     * A subclass must specify how to process the duplicate sets by overriding this method.
+     *
+     * @param readsWithSameUMI A set of reads with the matching UMIs with the same fragment start and end
+     * @param referenceContext A reference context object over the intervals determined by the duplicate set.
+     * @param featureContext Entries from a secondary feature file (e.g. vcf) if provided
+     *
+     */
+    public abstract void apply(ReadsWithSameUMI readsWithSameUMI, ReferenceContext referenceContext, FeatureContext featureContext );
+
+
+    /**
+     * Returns true for duplicate sets that does not meet required criteria for further processing.
+     * We encourage the user to override this method to meet their needs.
+     */
+    protected boolean rejectSet(final ReadsWithSameUMI readsWithSameUMI){
+        // Check that the set contains the minimum required number of reads in each strand
+        final Pair<Integer, Integer> strandCounts = MoleculeID.countStrands(readsWithSameUMI.getReads());
+        if (Math.min(strandCounts.getLeft(), strandCounts.getRight()) < minimumRequiredReadsPerStrand){
+            return true;
+        }
+
+        // Check that the read set is paired (this check may not reject some sets that contain unpaired reads)
```

Comment:
Is there a todo to make this smarter in the future or is % a good enough check?

Reply:
I think this is good enough for now